### PR TITLE
(fix) - O3-3189 - fix server-side pagination hook to handle error con…

### PR DESCRIPTION
**Pull request by chibongho**
…ditions; add hook to fetch all data from server-side paginated endpoints

# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [ ] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [x] My work includes tests or is validated by existing tests.
- [x] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

There were a few problems with this [PR](https://github.com/openmrs/openmrs-esm-core/pull/1114), which introduced new react hooks to handle server-side pagination. This new PR addresses them:
- The hooks provided could be used to fetch all results from the server (by requesting all pages), but it required some boilerplate code and was somewhat error prone (especially when handling request errors). This PR adds new hooks `useOpenmrsFetchAll` and `useFhirFetchAll` to address that.
- the `loadMore()` function returned by `useOpenmrsInfinite()` can result in infinite request loop as it did not handle request errors correctly.
- The hooks only provided support for OpenMRS REST endpoints. This PR also adds corresponding hooks to Fhir REST endpoints.

See corresponding [PR](https://github.com/openmrs/openmrs-esm-patient-management/pull/1292) (including video capture) in `esm-ward-app` for usages of these APIs.

### CODE REVIEW

File: packages/framework/esm-framework/docs/API.md [hunk 1]
```diff
@@ -6531,6 +6620,70 @@ ___
...
 
+### useFhirPagination
+
+▸ **useFhirPagination**<`T`\>(`url`, `pageSize`, `options?`): `Object`
+
+Fhir REST endpoints that return a list of objects, are server-side paginated.
+The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+if its size exceeds this limit.
+The max number of results per request is configurable server-side
+with the key "webservices.rest.maxResultsDefault". See: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25469882/REST+Module
```
Comment (mogoodrich) [inline packages/framework/esm-framework/docs/API.md:6631]: I'm not sure if the FHIR module uses this max result default as well, do you know @ibacher ?

File: packages/framework/esm-framework/docs/API.md [hunk 2]
```diff
@@ -5739,6 +5744,70 @@ ___
 
 ___
 
+### useFhirFetchAll
+
+▸ **useFhirFetchAll**<`T`\>(`url`, `options?`): `UseServerInfiniteReturnObject`<`T`, `fhir.Bundle`\>
+
+This hook handles fetching results from all pages of a paginated FHIR REST endpoint.
```
Comment (mogoodrich) [inline packages/framework/esm-framework/docs/API.md:5751]: Should/can we add more detail here?

File: packages/framework/esm-framework/docs/API.md [hunk 3]
```diff
@@ -5739,6 +5744,70 @@ ___
...
+
+#### Defined in
+
+[packages/framework/esm-react-utils/src/useFhirFetchAll.ts:18](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-react-utils/src/useFhirFetchAll.ts#L18)
+
+___
+
+### useFhirInfinite
+
+▸ **useFhirInfinite**<`T`\>(`url`, `options?`): `UseServerInfiniteReturnObject`<`T`, `fhir.Bundle`\>
```
Comment (mogoodrich) [inline packages/framework/esm-framework/docs/API.md:5786]: Should we/can we add more detail here?

File: packages/framework/esm-framework/docs/API.md [hunk 4]
```diff
@@ -6572,16 +6725,12 @@ ___
...
+▸ **useOpenmrsFetchAll**<`T`\>(`url`, `options?`): `UseServerInfiniteReturnObject`<`T`, `OpenMRSPaginatedResponse`<`T`\>\>
 
-Use this hook to paginate data that already exists on the client side.
-Note that if the data is obtained from server-side, the caller must handle server-side pagination manually.
-
-**`see`** `useServerPagination` for hook that automatically manages server-side pagination.
-
-**`see`** `useServerInfinite` for hook to get all data loaded onto the client-side
+Most OpenMRS REST endpoints that return a list of objects, such as getAll or search, are server-side paginated.
+This hook handles fetching results from all pages of a paginated OpenMRS REST endpoint.
```
Comment (mogoodrich) [inline packages/framework/esm-framework/docs/API.md:6733]: Can we better explain how his differs from useOpenMRSInfinite?

File: packages/framework/esm-framework/docs/interfaces/UseServerFetchAllOptions.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+- [fetcher](UseServerFetchAllOptions.md#fetcher)
+
+## UI Properties
+
+### partialData
+
+• `Optional` **partialData**: `boolean`
+
+If true, the data of any page as soon as they are fetched.
```
Comment (mogoodrich) [inline packages/framework/esm-framework/docs/interfaces/UseServerFetchAllOptions.md:34]: typo? missing verb?

File: packages/framework/esm-react-utils/src/useFhirPagination.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
+/** @module @category UI */
+import { type FetchResponse, makeUrl, openmrsFetch } from '@openmrs/esm-api';
+import { type ServerPaginationHandlers, useServerPagination, type UseServerPaginationOptions } from './useOpenmrsPagination';
+
+/**
+ * Fhir REST endpoints that return a list of objects, are server-side paginated.
+ * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+ * if its size exceeds this limit.
+ * The max number of results per request is configurable server-side
+ * with the key "webservices.rest.maxResultsDefault". See: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25469882/REST+Module
```
Comment (mogoodrich) [inline packages/framework/esm-react-utils/src/useFhirPagination.ts:10]: See my comment above

**Review by mogoodrich — CHANGES_REQUESTED** [review]
@chibongho  thanks for all your work on this...  I defer to @ibacher  @denniskigen  and @brandones  on the details, due to my lack of expertise in this area.

But, overall, this seems very well organized.  I did add a couple of comments.  One thing I didn't quite undersetand was the difference between "fetchAll" and "infinite".

Thanks again!

**Comment by chibongho** [comment]
The documentations are definitely on the messier side as is. I'll clean those up.

File: packages/framework/esm-framework/docs/API.md (hunk 1 above)
Comment (chibongho) [reply packages/framework/esm-framework/docs/API.md:6631]: I don't know what the config param is, but I think it's difficult. I'll just remove that line.

**Comment by chibongho** [comment]
@ibacher @brandones @denniskigen @vasharma05 This should be ready for review now. Thanks.

File: packages/framework/esm-react-utils/src/useFhirInfinite.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  type UseServerInfiniteOptions,
+  type UseServerInfiniteReturnObject,
+} from './useOpenmrsInfinite';
+
+/**
+ * Fhir REST endpoints that return a list of objects, are server-side paginated.
+ * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+ * if its size exceeds this limit.
+ *
+ * This function is the FHIR counterpart of `useOpenmrsInfinite`.
```
Comment (mogoodrich) [inline packages/framework/esm-react-utils/src/useFhirInfinite.ts:13]: This is the exact same description as for useFhirPagination?

File: packages/framework/esm-react-utils/src/useFhirPagination.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  useServerPagination,
+  type UseServerPaginationOptions,
+} from './useOpenmrsPagination';
+
+/**
+ * Fhir REST endpoints that return a list of objects, are server-side paginated.
+ * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+ * if its size exceeds this limit.
+ *
+ * This function is the FHIR counterpart of `useOpenmrsPagination`.
```
Comment (mogoodrich) [inline packages/framework/esm-react-utils/src/useFhirPagination.ts:14]: This is the exact same description as for useFhirInfinite?

File: packages/framework/esm-react-utils/src/useOpenmrsPagination.ts [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+ * Most OpenMRS REST endpoints that return a list of objects, such as getAll or search, are server-side paginated.
+ * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+ * if its size exceeds this limit.
+ * The max number of results per request is configurable server-side
+ * with the key "webservices.rest.maxResultsDefault". See: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25469882/REST+Module
+ *
+ * For any UI that displays a paginated view of the full data set, we MUST handle the server-side pagination properly,
+ * or else the UI does not correctly display the full data set.
+ * This hook does that by providing callback functions for navigating to different pages of the results, and
+ * lazy-loads the data on each page as needed.
```
Comment (mogoodrich) [inline packages/framework/esm-react-utils/src/useOpenmrsPagination.ts:40]: So is the different between useOpenmrsPagination and useOpenmrsInfinite just about the format of the callback functions it provides?

File: packages/framework/esm-react-utils/src/useOpenmrsInfinite.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Most REST endpoints that return a list of objects, such as getAll or search, are server-side paginated.
+ * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
+ * if its size exceeds this limit.
+ * The max number of results per request is configurable server-side
+ * with the key "webservices.rest.maxResultsDefault". See: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25469882/REST+Module
+ *
+ * This hook fetches data from a paginated rest endpoint, initially by fetching the first page of the results.
+ * It provides a callback to load data from subsequent pages as needed. This hook is intended to serve UIs that
+ * provide infinite loading / scrolling of results.
```
Comment (mogoodrich) [inline packages/framework/esm-react-utils/src/useOpenmrsInfinite.ts:84]: See my comment about useOpenmrsPagination...

**Review by mogoodrich — COMMENTED** [review]
Sorry, maybe it's me and I'm just not getting it, but I still don't understand the different between "Infinite" and "Pagination" in the two use methods...?

**Comment by chibongho** [comment]
> Sorry, maybe it's me and I'm just not getting it, but I still don't understand the different between "Infinite" and "Pagination" in the two use methods...?

(I find the useSWR documentation confusing on this as well, although we can certainly document our version of it better) If I understand correctly , `useInfinite` allows you do set `size` (the param for the # of pages to load) and it loads pages `0..size`. What it doesn't let you do is skipping to any arbitrary page without loading all pages preceding it, so if you only want to load the page 17, there is no way to do so without loading pages 0 to 16. 

`usePagination`, OTOH, allows skipping to any arbitrary page.

I'll add that do the comments.

File: packages/framework/esm-react-utils/src/useFhirInfinite.ts (hunk 7 above)
Comment (chibongho) [reply packages/framework/esm-react-utils/src/useFhirInfinite.ts:13]: The docs for `useOpenmrsPagination` and `useOpenmrsInfinite` are different though. Do you think that's ok?

**Comment by mogoodrich** [comment]
> (I find the useSWR documentation confusing on this as well, although we can certainly document our version of it better) If I understand correctly , `useInfinite` allows you do set `size` (the param for the # of pages to load) and it loads pages `0..size`. What it doesn't let you do is skipping to any arbitrary page without loading all pages preceding it, so if you only want to load the page 17, there is no way to do so without also loading pages 0 to 16.

If useSWR is confusing, fair.... in which case you could just (or also) point to the useSWR documentation for the *Pagination and *Infinite methods?

**Review by mogoodrich — APPROVED** [review]

File: packages/framework/esm-react-utils/src/useFhirInfinite.ts (hunk 7 above)
Comment (mogoodrich) [reply packages/framework/esm-react-utils/src/useFhirInfinite.ts:13]: Fair enough...

**Pull request merged**